unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/WebRequest/WebRequestController.cs

```diff
@@ -51,12 +53,164 @@ public static WebRequestController Create()
+            string url,
+            Hash128 hash,
+            Action<UnityWebRequest> onSuccess = null,
+            Action<UnityWebRequest> onfail = null,
+            int requestAttemps = 3,
+            int timeout = 0,
+            CancellationToken cancellationToken = default)
+        {
+            assetBundleFactory.SetHash(hash);
+            return await SendWebRequest(assetBundleFactory, url, null, onSuccess, onfail, requestAttemps,
+                timeout, cancellationToken);
+        }
+
+        public async UniTask<UnityWebRequest> GetTextureAsync(
+            string url,
+            Action<UnityWebRequest> onSuccess = null,
+            Action<UnityWebRequest> onfail = null,
+            int requestAttemps = 3,
+            int timeout = 0,
+            bool isReadable = true,
+            CancellationToken cancellationToken = default,
+            Dictionary<string, string> headers = null)
+        {
+            textureFactory.isReadable = isReadable;
+            return await SendWebRequest(textureFactory, url, null, onSuccess, onfail, requestAttemps,
+                timeout, cancellationToken, headers);
+        }
+
+        public async UniTask<UnityWebRequest> GetAudioClipAsync(
+            string url,
+            AudioType audioType,
+            Action<UnityWebRequest> onSuccess = null,
+            Action<UnityWebRequest> onfail = null,
+            int requestAttemps = 3,
+            int timeout = 0,
+            CancellationToken cancellationToken = default)
+        {
+            audioClipWebRequestFactory.SetAudioType(audioType);
+            return await SendWebRequest(audioClipWebRequestFactory, url, null, onSuccess, onfail, requestAttemps,
+                timeout, cancellationToken);
+        }
+        
+
+        private async UniTask<UnityWebRequest> SendWebRequest<T>(
+            T requestFactory,
+            string url,
+            DownloadHandler downloadHandler,
+            Action<UnityWebRequest> onSuccess,
+            Action<UnityWebRequest> onFail,
+            int requestAttemps,
+            int timeout,
+            CancellationToken cancellationToken,
+            Dictionary<string, string> headers = null) where T : IWebRequestFactory
+        {
+            requestAttemps = Mathf.Max(1, requestAttemps);
+            for (int i = requestAttemps - 1; i >= 0; i--)
+            {
+                UnityWebRequest request = requestFactory.CreateWebRequest(url);
+                request.timeout = timeout;
+
+                if (headers != null)
+                {
+                    foreach (var item in headers)
+                        request.SetRequestHeader(item.Key, item.Value);
+                }
+
+                if (downloadHandler != null)
+                    request.downloadHandler = downloadHandler;
+
+                UnityWebRequestAsyncOperation asyncOp = request.SendWebRequest();
+                while (!asyncOp.isDone)
```

Comment:
You can replace the loop with `await asyncOp.WithCancellation`

Reply:
Agree, feels more natural, will be adding those changes